File: backend/stability/circuit_breaker.go
package stability

import (
	"fmt"
	"sync"
	"time"
)

// CircuitState represents the state of a circuit breaker
type CircuitState string

const (
	StateClosed   CircuitState = "closed"   // Normal operation, requests allowed
	StateOpen     CircuitState = "open"     // Failed, requests blocked
	StateHalfOpen CircuitState = "half-open" // Testing if service recovered
)

// CircuitBreaker implements the circuit breaker pattern to prevent cascading failures
type CircuitBreaker struct {
	name              string
	failureThreshold  int           // Number of failures before opening circuit
	successThreshold  int           // Number of successes before closing circuit
	timeout           time.Duration // How long circuit stays open before trying again
	mu                sync.RWMutex
	state             CircuitState
	failureCount      int
	successCount      int
	lastFailureTime   time.Time
	lastStateChange   time.Time
}

// NewCircuitBreaker creates a new circuit breaker with default settings
// failureThreshold: 5 failures opens the circuit
// timeout: 60 seconds before attempting half-open
func NewCircuitBreaker(name string) *CircuitBreaker {
	return &CircuitBreaker{
		name:             name,
		failureThreshold: 5,
		successThreshold: 2,
		timeout:          60 * time.Second,
		state:            StateClosed,
		lastStateChange:  time.Now(),
	}
}

// RecordSuccess records a successful request
func (cb *CircuitBreaker) RecordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	cb.failureCount = 0

	if cb.state == StateHalfOpen {
		cb.successCount++
		if cb.successCount >= cb.successThreshold {
			cb.transitionToState(StateClosed)
			cb.successCount = 0
		}
	}
}

// RecordFailure records a failed request
func (cb *CircuitBreaker) RecordFailure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	cb.failureCount++
	cb.lastFailureTime = time.Now()
	cb.successCount = 0

	if cb.failureCount >= cb.failureThreshold && cb.state == StateClosed {
		cb.transitionToState(StateOpen)
	}
}

// CanExecute returns true if request can be executed
func (cb *CircuitBreaker) CanExecute() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case StateClosed:
		return true
	case StateOpen:
		// Check if timeout has passed, move to half-open
		if time.Since(cb.lastFailureTime) >= cb.timeout {
			cb.transitionToState(StateHalfOpen)
			cb.successCount = 0
			return true
		}
		return false
	case StateHalfOpen:
		return true
	}
	return false
}

// GetState returns current circuit breaker state
func (cb *CircuitBreaker) GetState() CircuitState {
	cb.mu.RLock()
	defer cb.mu.RUnlock()
	return cb.state
}

// GetStatus returns human-readable status
func (cb *CircuitBreaker) GetStatus() string {
	cb.mu.RLock()
	defer cb.mu.RUnlock()

	return fmt.Sprintf("state=%s failures=%d successes=%d changed=%v",
		cb.state, cb.failureCount, cb.successCount, cb.lastStateChange)
}

// transitionToState is internal method that changes state and logs it
func (cb *CircuitBreaker) transitionToState(newState CircuitState) {
	if cb.state != newState {
		oldState := cb.state
		cb.state = newState
		cb.lastStateChange = time.Now()
		fmt.Printf("[CircuitBreaker] %s: %s -> %s\n", cb.name, oldState, newState)
	}
}

// CircuitBreakerManager manages multiple circuit breakers for different datasources
type CircuitBreakerManager struct {
	mu          sync.RWMutex
	breakers    map[string]*CircuitBreaker
}

// NewCircuitBreakerManager creates a new manager
func NewCircuitBreakerManager() *CircuitBreakerManager {
	return &CircuitBreakerManager{
		breakers: make(map[string]*CircuitBreaker),
	}
}

// GetOrCreate gets or creates a circuit breaker for a datasource
func (cbm *CircuitBreakerManager) GetOrCreate(datasource string) *CircuitBreaker {
	cbm.mu.Lock()
	defer cbm.mu.Unlock()

	if breaker, exists := cbm.breakers[datasource]; exists {
		return breaker
	}

	breaker := NewCircuitBreaker(datasource)
	cbm.breakers[datasource] = breaker
	return breaker
}

// GetStatus returns status of all circuit breakers
func (cbm *CircuitBreakerManager) GetStatus() map[string]string {
	cbm.mu.RLock()
	defer cbm.mu.RUnlock()

	status := make(map[string]string)
	for name, breaker := range cbm.breakers {
		status[name] = breaker.GetStatus()
	}
	return status
}

// GetHealthStatus returns which datasources are healthy (circuit closed or half-open)
func (cbm *CircuitBreakerManager) GetHealthStatus() map[string]bool {
	cbm.mu.RLock()
	defer cbm.mu.RUnlock()

	health := make(map[string]bool)
	for name, breaker := range cbm.breakers {
		state := breaker.GetState()
		health[name] = state == StateClosed || state == StateHalfOpen
	}
	return health
}

File: src/utils/circuit-breaker.ts
/**
 * Circuit Breaker Pattern Implementation
 * Prevents cascading failures by failing fast when a service is unavailable
 */

export type CircuitState = 'closed' | 'open' | 'half-open';

export interface CircuitBreakerConfig {
  failureThreshold?: number; // Number of failures before opening (default: 5)
  timeout?: number; // Milliseconds to wait before half-open (default: 60000)
}

export class CircuitBreaker {
  private state: CircuitState = 'closed';
  private failureCount = 0;
  private failureThreshold: number;
  private timeout: number;
  private lastFailureTime: number | null = null;
  private successCount = 0;

  constructor(config: CircuitBreakerConfig = {}) {
    this.failureThreshold = config.failureThreshold || 5;
    this.timeout = config.timeout || 60000; // 60 seconds
  }

  /**
   * Check if the circuit allows execution
   */
  canExecute(): boolean {
    if (this.state === 'closed') {
      return true;
    }

    if (this.state === 'open') {
      // Check if timeout has elapsed
      if (this.lastFailureTime && Date.now() - this.lastFailureTime >= this.timeout) {
        this.state = 'half-open';
        this.successCount = 0;
        console.log('[CircuitBreaker] Transitioning to half-open state');
        return true;
      }
      return false;
    }

    // Half-open state
    return true;
  }

  /**
   * Record a successful call
   */
  recordSuccess(): void {
    this.failureCount = 0;

    if (this.state === 'half-open') {
      this.successCount++;
      // Allow 2 successful calls in half-open before closing
      if (this.successCount >= 2) {
        this.state = 'closed';
        console.log('[CircuitBreaker] Circuit closed - service recovered');
      }
    }
  }

  /**
   * Record a failed call
   */
  recordFailure(): void {
    this.failureCount++;
    this.lastFailureTime = Date.now();

    if (this.state === 'half-open') {
      // Immediately reopen on first failure in half-open
      this.state = 'open';
      console.log('[CircuitBreaker] Half-open failure - reopening circuit');
      return;
    }

    if (this.failureCount >= this.failureThreshold && this.state === 'closed') {
      this.state = 'open';
      console.log(
        `[CircuitBreaker] Failure threshold reached (${this.failureCount}/${this.failureThreshold}) - opening circuit`
      );
    }
  }

  /**
   * Get current circuit state
   */
  getState(): CircuitState {
    return this.state;
  }

  /**
   * Get failure count
   */
  getFailureCount(): number {
    return this.failureCount;
  }

  /**
   * Reset circuit to closed state
   */
  reset(): void {
    this.state = 'closed';
    this.failureCount = 0;
    this.successCount = 0;
    this.lastFailureTime = null;
  }
}

/**
 * Circuit Breaker Manager for handling multiple circuits
 * One circuit per data source (Prometheus, Loki, Database, etc)
 */
export class CircuitBreakerManager {
  private breakers = new Map<string, CircuitBreaker>();
  private readonly config: CircuitBreakerConfig;

  constructor(config: CircuitBreakerConfig = {}) {
    this.config = config;
  }

  /**
   * Get or create a circuit breaker for a service
   */
  getBreaker(serviceName: string): CircuitBreaker {
    if (!this.breakers.has(serviceName)) {
      this.breakers.set(serviceName, new CircuitBreaker(this.config));
    }
    return this.breakers.get(serviceName)!;
  }

  /**
   * Check if a service can be called
   */
  canExecute(serviceName: string): boolean {
    return this.getBreaker(serviceName).canExecute();
  }

  /**
   * Record success for a service
   */
  recordSuccess(serviceName: string): void {
    this.getBreaker(serviceName).recordSuccess();
  }

  /**
   * Record failure for a service
   */
  recordFailure(serviceName: string): void {
    this.getBreaker(serviceName).recordFailure();
  }

  /**
   * Get all breaker states
   */
  getAllStates(): Record<string, CircuitState> {
    const states: Record<string, CircuitState> = {};
    this.breakers.forEach((breaker, name) => {
      states[name] = breaker.getState();
    });
    return states;
  }

  /**
   * Reset all circuits
   */
  resetAll(): void {
    this.breakers.forEach((breaker) => {
      breaker.reset();
    });
  }

  /**
   * Reset a specific service's circuit
   */
  reset(serviceName: string): void {
    this.breakers.get(serviceName)?.reset();
  }
}

// Global circuit breaker manager instance
export const circuitBreakerManager = new CircuitBreakerManager({
  failureThreshold: 5,
  timeout: 60000, // 60 seconds
});
